src/main/java/ai/promoted/delivery/client/InsertionFactory.java

```diff
@@ -16,10 +19,50 @@ public class InsertionFactory {
    * @return a populated Insertion
    */
   public static Insertion createInsertionWithProperties(String contentId, Map<String, Object> properties) {
-    Properties props = new Properties().structField(properties);
-    return new Insertion().contentId(contentId).properties(props);
+    Insertion ins = new Insertion().contentId(contentId);
+    if (!properties.isEmpty()) {
+      ins.properties(Properties.newBuilder().setStruct(convertMapToStruct(properties)).build());
+    }
+    return ins;
+  }
+
+  private static com.google.protobuf.Struct convertMapToStruct(Map<String, Object> map) {
+    com.google.protobuf.Struct.Builder structBuilder = com.google.protobuf.Struct.newBuilder();
+    map.forEach((key, value) -> {
+      com.google.protobuf.Value protoValue = convertObjectToValue(value);
+      structBuilder.putFields(key, protoValue);
+    });
+    return structBuilder.build();
   }
-  
+
+  private static com.google.protobuf.Value convertObjectToValue(Object object) {
+    if (object instanceof Boolean) {
+      return com.google.protobuf.util.Values.of((Boolean) object);
+    } else if (object instanceof Integer) {
+      return com.google.protobuf.util.Values.of((Integer) object);
+    } else if (object instanceof Long) {
+      return com.google.protobuf.util.Values.of((Long) object);
+    } else if (object instanceof Float) {
+      return com.google.protobuf.util.Values.of((Float) object);
+    } else if (object instanceof Double) {
+      return com.google.protobuf.util.Values.of((Double) object);
+    } else if (object instanceof String) {
+      return com.google.protobuf.util.Values.of((String) object);
+    } else if (object instanceof Iterable) {
+      Iterable<?> iterable = (Iterable<?>) object;
+      ListValue.Builder listBuilder = ListValue.newBuilder();
+      for (Object item : iterable) {
+        listBuilder.addValues(convertObjectToValue(item));
+      }
+      return com.google.protobuf.util.Values.of(listBuilder.build());
+    } else if (object instanceof Map) {
+      @SuppressWarnings("unchecked")
+      Map<String, Object> map = (Map<String, Object>) object;
+      return com.google.protobuf.util.Values.of(convertMapToStruct(map));
+    }
+    return com.google.protobuf.Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
```

Comment:
@prm-dan Do you have a preference on how this is handled? Like if this should be logging an error, throwing some kind of unimplemented exception (and if so, do we catch it inside the SDK or document it in the README), etc.

Reply:
Actually, this isn't even used in the SDK (outside of test). So whatever we do here could be worth documenting

Reply:
Hmm, I see.  I'd probably throw if it's an unsupported type.  They'd hit it when developing.  You'll also want an explicit null check.

```
} else if (object == null) {
  return com.google.protobuf.Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
} else {
  throw new UnsupportedOperationException("property value class=" + object.getClass());
}
```

Reply:
Done.